E-Commerce Sales Data Analysis

PART 1 - Data Preprocessing:
- Loading the Dataset
- Cleaning the Dataset
- Data Transformation

In [91]:
import pandas as pd
import numpy as np
from pandas import read_csv
import matplotlib.pyplot as plt

Loading and Analyzing the Raw Data

In [92]:
df = pd.read_csv('/home/user/Desktop/E-commerce Data-Analysis/amazon.csv')
df

product_id                                       product_name  \
0     B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1     B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2     B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3     B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4     B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   
...          ...                                                ...   
1460  B08L7J3T31  Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...   
1461  B01M6453MB  Prestige Delight PRWO Electric Rice Cooker (1 ...   
1462  B009P2LIL4  Bajaj Majesty RX10 2000 Watts Heat Convector R...   
1463  B00J5DYCCA  Havells Ventil Air DSP 230mm Exhaust Fan (Pist...   
1464  B01486F4G6  Borosil Jumbo 1000-Watt Grill Sandwich Maker (...   

                                               category discounted_price  \
0     Computers&Accessories|Accessories&Peripherals|...             ₹399   
1     Computers&Accessories|Accessories&Peripherals|...             ₹199   
2     Computers&Accessories|Accessories&Peripherals|...             ₹199   
3     Computers&Accessories|Accessories&Peripherals|...             ₹329   
4     Computers&Accessories|Accessories&Peripherals|...             ₹154   
...                                                 ...              ...   
1460  Home&Kitchen|Kitchen&HomeAppliances|WaterPurif...             ₹379   
1461  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...           ₹2,280   
1462  Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...           ₹2,219   
1463  Home&Kitchen|Heating,Cooling&AirQuality|Fans|E...           ₹1,399   
1464  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...           ₹2,863   

     actual_price discount_percentage rating rating_count  \
0          ₹1,099                 64%    4.2       24,269   
1            ₹349                 43%    4.0       43,994   
2          ₹1,899                 90%    3.9        7,928   
3            ₹699                 53%    4.2       94,363   
4            ₹399                 61%    4.2       16,905   
...           ...                 ...    ...          ...   
1460         ₹919                 59%      4        1,090   
1461       ₹3,045                 25%    4.1        4,118   
1462       ₹3,080                 28%    3.6          468   
1463       ₹1,890                 26%      4        8,031   
1464       ₹3,690                 22%    4.3        6,987   

                                          about_product  \
0     High Compatibility : Compatible With iPhone 12...   
1     Compatible with all Type C enabled devices, be...   
2     【 Fast Charger& Data Sync】-With built-in safet...   
3     The boAt Deuce USB 300 2 in 1 cable is compati...   
4     [CHARGE & SYNC FUNCTION]- This cable comes wit...   
...                                                 ...   
1460  SUPREME QUALITY 90 GRAM 3 LAYER THIK PP SPUN F...   
1461                       230 Volts, 400 watts, 1 Year   
1462  International design and styling|Two heat sett...   
1463  Fan sweep area: 230 MM ; Noise level: (40 - 45...   
1464  Brand-Borosil, Specification – 23V ~ 5Hz;1 W...   

                                                user_id  \
0     AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1     AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2     AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3     AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4     AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   
...                                                 ...   
1460  AHITFY6AHALOFOHOZEOC6XBP4FEA,AFRABBODZJZQB6Z4U...   
1461  AFG5FM3NEMOL6BNFRV2NK5FNJCHQ,AGEINTRN6Z563RMLH...   
1462  AGVPWCMAHYQWJOQKMUJN4DW3KM5Q,AF4Q3E66MY4SR7YQZ...   
1463  AF2JQCLSCY3QJATWUNNHUSVUPNQQ,AFDMLUXC5LS5RXDJS...   
1464  AFGW5PT3R6ZAVQR4Y5MWVAKBZAYA,AG7QNJ2SCS5VS5VYY...   

                                              user_name  \
0     Manav,Adarsh gupta,Sunde

In [93]:
# Gives an Overview of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   category             1465 non-null   object
 3   discounted_price     1465 non-null   object
 4   actual_price         1465 non-null   object
 5   discount_percentage  1465 non-null   object
 6   rating               1465 non-null   object
 7   rating_count         1463 non-null   object
 8   about_product        1465 non-null   object
 9   user_id              1465 non-null   object
 10  user_name            1465 non-null   object
 11  review_id            1465 non-null   object
 12  review_title         1465 non-null   object
 13  review_content       1465 non-null   object
 14  img_link             1465 non-null   object
 15  product_link         1465 non-null   object
dtypes: obj

In [94]:
df.describe()

product_id                                       product_name  \
count         1465                                               1465   
unique        1351                                               1337   
top     B083342NKJ  Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...   
freq             3                                                  5   

                                                 category discounted_price  \
count                                                1465             1465   
unique                                                211              550   
top     Computers&Accessories|Accessories&Peripherals|...             ₹199   
freq                                                  233               53   

       actual_price discount_percentage rating rating_count  \
count          1465                1465   1465         1463   
unique          449                  92     28         1143   
top            ₹999                 50%    4.1        9,378   
freq            120                  56    244            9   

                                            about_product  \
count                                                1465   
unique                                               1293   
top     [CHARGE & SYNC FUNCTION]- This cable comes wit...   
freq                                                    6   

                                                  user_id  \
count                                                1465   
unique                                               1194   
top     AHIKJUDTVJ4T6DV6IUGFYZ5LXMPA,AE55KTFVNXYFD5FPY...   
freq                                                   10   

                                                user_name  \
count                                                1465   
unique                                               1194   
top     $@|\|TO$|-|,Sethu madhav,Akash Thakur,Burger P...   
freq                                                   10   

                                                review_id  \
count                                                1465   
unique                                               1194   
top     R3F4T5TRYPTMIG,R3DQIEC603E7AY,R1O4Z15FD40PV5,R...   
freq                                                   10   

                                             review_title  \
count                                                1465   
unique                                               1194   
top     Worked on iPhone 7 and didn’t work on XR,Good ...   
freq                                                   10   

                                           review_content  \
count                                                1465   
unique                                               1212   
top     I am not big on camera usage, personally. I wa...   
freq                                                    8   

                                                 img_link  \
count                                                1465   
unique                                               1412   
top     https://m.media-amazon.com/images/I/413sCRKobN...   
freq                                                    3   

                                             product_link  
count                                                1465  
unique                                               1465  
top     https://www.amazon.in/Borosil-Jumbo-1000-Watt-...  
freq                                                    1

Dataset Overview:

- Our Dataset has 16 columns and 1465 columns
- All Columns are stored as Object type, Including numeric data.
- rating_count Column has 2 missing values.
- product_name, about_product, review_title and review_content columns seem lengthy.
- Some users and reviews might be duplicated.
- Category Column has 211 unique values, meaning products belong to 211 categories.

In [95]:
# Filled the missing values with 0, I assumed that missing means no rating.
df['rating_count'] = df['rating_count'].fillna(0)

Cleaning the Dataset

In [96]:
# I converted the columns to strings to avoid errors when using the str function
# Removed unwanted symbols and converted them to float.
df['discounted_price'] = df['discounted_price'].astype(str).str.replace('₹', '').str.replace(',', '').astype(float)
df['actual_price'] = df['actual_price'].astype(str).str.replace('₹', '').str.replace(',', '').astype(float)
df['discount_percentage'] = df['discount_percentage'].astype(str).str.replace('%', '').astype(float)
df['rating'] = pd.to_numeric(df['rating'], errors='coerce')  # Convert rating safely
df['rating_count'] = df['rating_count'].astype(str).str.replace(',', '').astype(float)

In [97]:
# Verified the column types to ensure numerical columns were converted correctly.
print(df.dtypes)

product_id              object
product_name            object
category                object
discounted_price       float64
actual_price           float64
discount_percentage    float64
rating                 float64
rating_count           float64
about_product           object
user_id                 object
user_name               object
review_id               object
review_title            object
review_content          object
img_link                object
product_link            object
dtype: object


In [98]:
# Checked the length of text in long text columns to decide on how to shorten
df[['product_name', 'about_product', 'review_title', 'review_content']].applymap(lambda x: len(str(x)))

/tmp/ipykernel_36503/3850513996.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df[['product_name', 'about_product', 'review_title', 'review_content']].applymap(lambda x: len(str(x)))


product_name  about_product  review_title  review_content
0              162            960           124             483
1              193            605           181            1186
2              136           1091           137             271
3              199            699           140             443
4              110            574           138            2451
...            ...            ...           ...             ...
1460           183            423           154             479
1461            55             28           328            1036
1462            78            297           118             880
1463            54            528            95             388
1464            52            391           214            1457

[1465 rows x 4 columns]

In [99]:
# Trimmed the long texts keeping it short and readable.
df['product_name'] = df['product_name'].str.slice(0, 50)
df['about_product'] = df['about_product'].str.slice(0, 100)
df['review_title'] = df['review_title'].str.slice(0, 50)
df['review_content'] = df['review_content'].str.slice(0, 100)

In [100]:
# Count unique values in each column
# Most Columns have unique values, so no encoding needed.
df.nunique()

product_id             1351
product_name           1289
category                211
discounted_price        550
actual_price            449
discount_percentage      92
rating                   25
rating_count           1144
about_product          1190
user_id                1194
user_name              1194
review_id              1194
review_title           1187
review_content         1189
img_link               1412
product_link           1465
dtype: int64

In [101]:
# Checked for Duplicates
duplicates = df.duplicated().sum()
print(f"Total duplicate rows: {duplicates}")

Total duplicate rows: 0


In [102]:
# Check for missing values
# Rating Column has 1 missing value
missing_values = df.isnull().sum()
print(missing_values[missing_values > 0])  # Show only columns with missing values

rating    1
dtype: int64


In [103]:
# Dropped the row with the missing value
df = df.dropna(subset=['rating'])

In [ ]:
# Checking the first 10 rows of the Category Column before extracting main Categories
df['category'].head(10)

0    Computers&Accessories|Accessories&Peripherals|...
1    Computers&Accessories|Accessories&Peripherals|...
2    Computers&Accessories|Accessories&Peripherals|...
3    Computers&Accessories|Accessories&Peripherals|...
4    Computers&Accessories|Accessories&Peripherals|...
5    Computers&Accessories|Accessories&Peripherals|...
6    Computers&Accessories|Accessories&Peripherals|...
7    Computers&Accessories|Accessories&Peripherals|...
8    Computers&Accessories|NetworkingDevices|Networ...
9    Computers&Accessories|Accessories&Peripherals|...
Name: category, dtype: object

In [ ]:
# Extracting only the main category
df['category'] = df['category'].str.split('|').str[0]

# Counting unique values after extracting main categories
df['category'].value_counts()

/tmp/ipykernel_36503/3609899837.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['category'] = df['category'].str.split('|').str[0]


category
Electronics              526
Computers&Accessories    453
Home&Kitchen             447
OfficeProducts            31
MusicalInstruments         2
HomeImprovement            2
Toys&Games                 1
Car&Motorbike              1
Health&PersonalCare        1
Name: count, dtype: int64

In [105]:
# Created a new Column called discount_amount
df['discount_amount'] = df['actual_price'] - df['discounted_price']

/tmp/ipykernel_36503/1411443971.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['discount_amount'] = df['actual_price'] - df['discounted_price']


In [ ]:
# I checked random 10 rows to check for timestamps
# No timestamp data is available
df[['review_id', 'review_title', 'review_content']].sample(10)  


review_id  \
1176  R1BE774NJ5R2DX,R1U4G4C65P8D4G,R2WMQC1KWG94P7,R...   
641   R2ZYS8OJWNY7VY,R33U0ERE0GVMNJ,R1CQTXZAM4625F,R...   
353   R2A7MIUNOW8DOE,R2FXP703540FR1,R37E7QJET0BYE8,R...   
405   RU8SZ6NFWFYV6,R1GQJT5423OND1,R2OJEFG3PL2ZVW,R3...   
1193  RXAODV2OHBKW4,R2AV4UYNGRE33Q,R3KJCPWOGYC672,R2...   
214   R1H0YNK5FI6IM9,RRVOLO108F914,R18D45T6ZYK9SS,R9...   
439   R36T09OX35WPH0,R1SPKNBAZ5I7N1,R2H32V6C3AL47P,R...   
1372  R1HD4L4O8FYBVJ,R1DSP7AK9O1EG0,R3F6O9LJWF9UGP,R...   
643   R1LAI2YEEUW0E0,RR8Y3CSNEHCK6,R1MXV3ILO9VTIP,RJ...   
741   R1IF6OX5EMCHRA,R11D9Q1R128JAS,RYHX3VX289G1,R19...   

                                           review_title  \
1176  Good for the price,Good product,Good Deal,Defe...   
641   Got less storage.,Good deal,Looking is fake pr...   
353   Fake Product,Costly but excellent quality,Stor...   
405   Little above average Earbuds,Buds i love,Uniqu...   
1193  Compact and powerful but alot of plastic,Overa...   
214   Good compatibility,Good Product,Good nice serv...   
439   Budget Phone,Good product at this price,Good p...   
1372  Meets expectation,Good product,High quality pr...   
643   Best.,Price is high,Trusted and genuine HP ink...   
741   Good product but little costly,Very sturdy and...   

                                         review_content  
1176  Good one for usability,Its value for money. Co...  
641   I ordered 128GB but got 119GB.,It's a good dea...  
353   128GB SD Card is showing 134GBDon't buy this p...  
405   The earphones have good clear sound quality. I...  
1193  Firstly I got the cracked jar but the product ...  
214   The remote works well and is pleasantly compat...  
439   If your budget is low and you are looking for ...  
1372  An easy-to-use coffee filter at a reasonable p...  
643   Best Fingerprint Scanner Divice.,Easy to use,I...  
741   If you are using a seperate mouse go without c...